src/utils/viewer/viewer.cpp

```diff
@@ -381,79 +294,59 @@ void Viewer::addTemplateObject() {
 
 }  // addPrimitiveObject
 
 void Viewer::removeLastObject() {
-  if (objectIDs_.size() == 0) {
+  auto existingObjectIDs = simulator_->getExistingObjectIDs();
+  if (existingObjectIDs.size() == 0) {
     return;
   }
-  physicsManager_->removeObject(objectIDs_.back());
-  objectIDs_.pop_back();
+  simulator_->removeObject(existingObjectIDs.back());
 }
 
 void Viewer::invertGravity() {
-  const Mn::Vector3& gravity = physicsManager_->getGravity();
+  const Mn::Vector3& gravity = simulator_->getGravity();
   const Mn::Vector3 invGravity = -1 * gravity;
-  physicsManager_->setGravity(invGravity);
+  simulator_->setGravity(invGravity);
 }
 
 void Viewer::pokeLastObject() {
-  if (objectIDs_.size() == 0)
+  auto existingObjectIDs = simulator_->getExistingObjectIDs();
+  if (existingObjectIDs.size() == 0)
     return;
   Mn::Matrix4 T =
       agentBodyNode_->MagnumObject::transformationMatrix();  // Relative to
                                                              // agent bodynode
   Mn::Vector3 impulse = T.transformVector({0.0f, 0.0f, -3.0f});
   Mn::Vector3 rel_pos = Mn::Vector3(0.0f, 0.0f, 0.0f);
-  physicsManager_->applyImpulse(objectIDs_.back(), impulse, rel_pos);
+
+  simulator_->applyImpulse(impulse, rel_pos, existingObjectIDs.back());
 }
 
 void Viewer::pushLastObject() {
-  if (objectIDs_.size() == 0)
+  auto existingObjectIDs = simulator_->getExistingObjectIDs();
+  if (existingObjectIDs.size() == 0)
     return;
   Mn::Matrix4 T =
       agentBodyNode_->MagnumObject::transformationMatrix();  // Relative to
                                                              // agent bodynode
   Mn::Vector3 force = T.transformVector({0.0f, 0.0f, -40.0f});
   Mn::Vector3 rel_pos = Mn::Vector3(0.0f, 0.0f, 0.0f);
-  physicsManager_->applyForce(objectIDs_.back(), force, rel_pos);
-}
-
-void Viewer::recomputeNavMesh(const std::string& sceneFilename,
-                              esp::nav::NavMeshSettings& navMeshSettings) {
-  esp::nav::PathFinder::ptr pf = esp::nav::PathFinder::create();
-
-  esp::assets::MeshData::uptr joinedMesh =
-      resourceManager_.createJoinedCollisionMesh(sceneFilename);
-
-  if (!pf->build(navMeshSettings, *joinedMesh)) {
-    LOG(ERROR) << "Failed to build navmesh";
-    return;
-  }
-
-  LOG(INFO) << "reconstruct navmesh successful";
-  pathfinder_ = pf;
-
-  // reset the visualization if necessary
-  if (navMeshVisNode_ != nullptr) {
-    toggleNavMeshVisualization();  // first clear old vis
-    toggleNavMeshVisualization();
-  }
+  simulator_->applyForce(force, rel_pos, existingObjectIDs.back());
 }
 
 void Viewer::torqueLastObject() {
-  if (objectIDs_.size() == 0)
+  auto existingObjectIDs = simulator_->getExistingObjectIDs();
+  if (existingObjectIDs.size() == 0)
     return;
   Mn::Vector3 torque = randomDirection() * 30;
-  physicsManager_->applyTorque(objectIDs_.back(), torque);
+  simulator_->applyTorque(torque, existingObjectIDs.back());
```

Comment:
Interesting convention we have of torque, object_id, would expect it the other way around.

Reply:
We do it this way to later support set(thing, objectId, sceneId) with sceneId given a default value.